embedded/store/kvconstraints.go

```diff
@@ -0,0 +1,51 @@
+/*
+Copyright 2022 CodeNotary, Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package store
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/codenotary/immudb/embedded/tbtree"
+)
+
+type KVConstraints struct {
+	Key                []byte
```

Comment:
wouldn't expect the key to belong to a constraint datatype...

Reply:
it'd be great to have a constraint interface... so it may be extensible in the future i.e. mustExist, mustNoExist would be concrete implementations of such interface

Reply:
As discussed offline, I think each constraint should be a functor that takes the current db state and responds with true or false depending on whether the constraint is met or not. Such function could be parametrized, e.g. MustExist is parametrised with the key, same with MustNotExist, for NotModifiedAfterTX it would be parametrised with key and transaction id.

Let's maybe first focus on the GRPC protocol since that's something hard to change later.

@jeroiraz what do you think about this approach?:

```proto
message WriteConstraint {

	message KeyMustExist {
		bytes key = 1;
	}
	
	message KeyMustNotExist {
		bytes key = 1;
	}
	
	message KeyNotModifiedAfterTX {
		bytes key = 1;
		uint64 txID = 2;
	}

	oneof constraint {
		KeyMustExist keyMustExist = 1;
		KeyMustNotExist keyMustNotExist = 2;
		KeyNotModifiedAfterTX keyNotModifiedAfter = 3;
	}
}
```
and then in messages:
```proto
message SetRequest {
	repeated KeyValue KVs = 1;
	bool  noWait = 2;
	repeated WriteConstraint constraints = 3;
}
```

Internally we could represent this as an implementation of abstract interface so that it's easy to extend with more constraints such as those based on prefix (e.g. there must be no key with given prefix).